web/src/hotkey.js

```diff
@@ -163,107 +196,49 @@ const keydown_either_mappings = {
-    i: {name: "message_actions", message_view_only: true},
-    j: {name: "vim_down", message_view_only: true},
-    k: {name: "vim_up", message_view_only: true},
-    l: {name: "vim_right", message_view_only: true},
-    m: {name: "move_message", message_view_only: true},
-    n: {name: "n_key", message_view_only: false},
-    p: {name: "p_key", message_view_only: false},
-    q: {name: "query_streams", message_view_only: true},
-    r: {name: "reply_message", message_view_only: true},
-    s: {name: "toggle_conversation_view", message_view_only: true},
-    t: {name: "open_recent_view", message_view_only: true},
-    u: {name: "toggle_sender_info", message_view_only: true},
-    v: {name: "show_lightbox", message_view_only: true},
-    w: {name: "query_users", message_view_only: true},
-    x: {name: "compose_private_message", message_view_only: true},
-    z: {name: "zoom_to_message_near", message_view_only: true},
+    "Shift+Backspace": {name: "backspace", message_view_only: true},
+    "Shift+Enter": {name: "enter", message_view_only: false},
+    "Shift+Delete": {name: "delete", message_view_only: false},
 };
 
 // Keyboard Event Viewer tool:
 // https://w3c.github.io/uievents/tools/key-event-viewer.html
 export function get_keydown_hotkey(e) {
-    let hotkey;
-
-    if (e.altKey) {
-        hotkey = keydown_alt_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
+    if (common.has_mac_keyboard() && e.ctrlKey && e.key !== "[") {
+        // On macOS, Cmd is used instead of Ctrl. Except 'Ctrl + ['.
         return undefined;
     }
-
-    if (e.ctrlKey && !e.shiftKey) {
-        hotkey = keydown_ctrl_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
-    }
-
-    const isCmdOrCtrl = common.has_mac_keyboard() ? e.metaKey : e.ctrlKey;
-    if (isCmdOrCtrl && !e.shiftKey) {
-        hotkey = keydown_cmd_or_ctrl_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
-        return undefined;
-    } else if (e.metaKey || e.ctrlKey) {
+    if (!common.has_mac_keyboard() && e.metaKey) {
         return undefined;
     }
 
-    if (e.shiftKey) {
-        hotkey = keydown_shift_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
-    }
+    // Make lowercase a-z uppercase; leave others unchanged.
+    // We distinguish between 'a' and 'A' as 'A' vs 'Shift+A'.
+    // This helps to avoid Caps Lock affecting shortcuts.
+    let key = /^[a-z]$/.test(e.key) ? e.key.toUpperCase() : e.key;
 
-    if (!e.shiftKey) {
-        hotkey = keydown_unshift_mappings[e.key];
-        if (hotkey) {
-            return hotkey;
-        }
-    }
+    // We ignore the modifiers for these keys as they are treated as
+    // involved in generating the character in various keyboard layouts.
+    // Here, modifiers are not the part of the shortcut key combination.
+    const ignore_modifiers_for_keys = new Set(["*", "+", "=", "-", "/", ":", "<", ">", "?", "@"]);
```

Comment:
TODO: We should define this set outside the function.

Reply:
Yeah we can define this just before `KEYDOWN_MAPPINGS`, and call it `IGNORE_MODIFIERS_KEYS` with a comment along the lines of:

> This collection of characters are those where many keyboards require modifier keys in order to produce, or where pressing this key with a modifier key pressed is expected to produce a different key . We ignore all modifiers when processing these keys, and disallow assigning hotkeys to any of these.

Reply:
```js
// A set of characters that, on many keyboard layouts, require modifier keys
// to type, or where pressing the key with a modifier is expected to produce
// a different character. We ignore all modifiers when processing these keys,
// and disallow assigning hotkeys that involve any modifier combined with them.
const KNOWN_IGNORE_MODIFIERS_KEYS = new Set(["*", "+", "=", "-", "/", ":", "<", ">", "?", "@"]);
```
Updated the comment and moved it out of the function.

Also, added a prefix `KNOWN` to the variable name as it is not a complete list of such characters but the ones we use currently. -- Same reason as https://github.com/zulip/zulip/pull/34570#discussion_r2087649881